File: pset6/credit/credit.py
###

# credit.py

# M. Ashuza

# A program for validating credit card numbers.

###
from cs50 import get_int


def main():
    # Get the cc_number
    while True:
        cc_number = get_int("Number: ")
        if cc_number >= 0:
            break

    # Check validity of card brand
    if check_validity(cc_number):
        print_ccn_brand(cc_number)
    else:
        print("INVALID")


def check_validity(ccn):
    return checkSum(ccn)


def checkSum(ccn):
    sum = 0
    # Converting ccn into a string
    # 1037 -> "1037"
    # Get the value from 0 up to lenht of str -1
    for i in range(len(str(ccn))):
        if (i % 2 == 0):
            sum += ccn % 10
        else:
            # 2 * 9 = 18
            digit = 2 * (ccn % 10)
            # break that 18 as 1 + 8
            sum += digit // 10 + digit % 10
        ccn //= 10  # ccn needs to be reduced so that to move one digit over

    return sum % 10 == 0


# Print the card brand
def print_ccn_brand(ccn):
    # American express brand
    if (ccn >= 34e13 and ccn < 35e13) or (ccn >= 37e13 and ccn < 38e13):
        print("AMEX")
    elif ccn >= 51e14 and ccn < 56e14:
        print("MASTERCARD")
    elif (ccn >= 4e12 and ccn < 5e12) or (ccn >= 4e15 and ccn < 5e15):
        print("VISA")
    else:
        print("INVALID")


# Python syntax for runing the main function
if __name__ == "__main__":
    main()

File: C/pset1/credit/credit.c
/**

  * credit.c

  * M. Ashuza

  * A program for validating credit card numbers.

  **/

#include <cs50.h>
#include <math.h>
#include <stdio.h>

bool check_validity(long long cc_number);
bool checksum(long long cc_n);
void print_cc_number(long long cc_number);
int find_lenght(long long n);

int main(void)
{
    long long cc_number;
    // Get and validate user credit card number
    do
    {
        cc_number = get_long_long("Number: ");
    }
    while (cc_number < 0);

    // Check validity of the card brand
    if (check_validity(cc_number))
    {
        print_cc_number(cc_number);
    }
    else
    {
        printf("INVALID\n");
    }
}

bool check_validity(long long cc_number)
{
    // check the length brand
    int len = find_lenght(cc_number);
    return (len == 13 || len == 15 || len == 16) && checksum(cc_number);
}

// Get the lenght of the ccnumber
int find_lenght(long long n)
{
    int len;
    for (len = 0; n != 0; n /= 10)
    {
        len++;
    }
    return len;
}

// Luhn’s Algorithm /check sum

bool checksum(long long cc_n)
{
    int sum = 0;
    for (int i = 0; cc_n != 0; i++, cc_n /= 10)
    {
        if (i % 2 == 0)
        {
            sum += cc_n % 10; 
        }
        else
        {
            int digit = 2 * (cc_n % 10); // Multiply every other digit by 2
            sum += digit / 10 + digit % 10;
        }
    }
    return (sum % 10) == 0;
}

// Print the brand
void print_cc_number(long long cc_n)
{
    if ((cc_n >= 34e13 && cc_n < 35e13) || (cc_n >= 37e13 && cc_n < 38e13))
    {
        printf("AMEX\n");
    }
    else if (cc_n >= 51e14 && cc_n < 56e14)
    {
        printf("MASTERCARD\n");
    }
    else if ((cc_n >= 4e12 && cc_n < 5e12) || (cc_n >= 4e15 && cc_n < 5e15))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}



